src/vos/vos_layout.h

```diff
@@ -363,16 +363,20 @@ struct vos_irec_df {
  */
 struct vos_obj_df {
 	daos_unit_oid_t			vo_id;
-	/** The latest sync epoch */
+	/** The latest sync epoch, is DAOS_EPOCH_MAX for flattened object */
 	daos_epoch_t			vo_sync;
 	/** Offset of known existing dkey */
 	umem_off_t			vo_known_dkey;
 	/** Attributes for future use */
 	daos_epoch_t			vo_max_write;
 	/** Incarnation log for the object */
 	struct ilog_df			vo_ilog;
-	/** VOS dkey btree root */
-	struct btr_root			vo_tree;
+	union {
+		/** VOS dkey btree root for hierarchical object */
+		struct btr_root		vo_tree;
+		/** external address for flattened object */
+		bio_addr_t		vo_addr;
```

Comment:
D_CASSERT(sizeof(bio_addr_t) <= sizeof(struct bio_root));

Reply:
sure, will add when refresh this patch or in next patch, thx